kyo-core/shared/src/main/scala/kyo/Timer.scala

```diff
@@ -1,288 +1,87 @@
-            final private class FutureTimerTask(task: ScheduledFuture[?]) extends TimerTask.Unsafe:
-                def cancel()(using AllowUnsafe): Boolean    = task.cancel(false)
-                def cancelled()(using AllowUnsafe): Boolean = task.isCancelled()
-                def done()(using AllowUnsafe): Boolean      = task.isDone()
-            end FutureTimerTask
-
-            def schedule(delay: Duration)(f: => Unit)(using AllowUnsafe): TimerTask.Unsafe =
-                if delay.isFinite then
-                    val call = new Callable[Unit]:
-                        def call: Unit = f
-                    new FutureTimerTask(exec.schedule(call, delay.toNanos, TimeUnit.NANOSECONDS))
-                else
-                    TimerTask.Unsafe.noop
-
-            def scheduleAtFixedRate(
-                initialDelay: Duration,
-                period: Duration
-            )(f: => Unit)(using AllowUnsafe): TimerTask.Unsafe =
-                if period.isFinite && initialDelay.isFinite then
-                    new FutureTimerTask(
-                        exec.scheduleAtFixedRate(
-                            () => f,
-                            initialDelay.toNanos,
-                            period.toNanos,
-                            TimeUnit.NANOSECONDS
-                        )
-                    )
-                else
-                    TimerTask.Unsafe.noop
-
-            def scheduleWithFixedDelay(
-                initialDelay: Duration,
-                period: Duration
-            )(f: => Unit)(using AllowUnsafe): TimerTask.Unsafe =
-                if period.isFinite && initialDelay.isFinite then
-                    new FutureTimerTask(
-                        exec.scheduleWithFixedDelay(
-                            () => f,
-                            initialDelay.toNanos,
-                            period.toNanos,
-                            TimeUnit.NANOSECONDS
-                        )
-                    )
-                else
-                    TimerTask.Unsafe.noop
-    end Unsafe
+    def repeatWithDelay[E, S](delay: Duration)(f: => Unit < (Async & Abort[E]))(using Frame): Fiber[E, Unit] < IO =
+        repeatWithDelay(Duration.Zero, delay)(f)
+
+    def repeatWithDelay[E, S](
+        startAfter: Duration,
+        delay: Duration
+    )(
+        f: => Unit < (Async & Abort[E])
+    )(using Frame): Fiber[E, Unit] < IO =
+        repeatWithDelay(startAfter, delay, ())(_ => f)
+
+    def repeatWithDelay[E, A: Flat, S](
+        startAfter: Duration,
+        delay: Duration,
+        state: A
+    )(
+        f: A => A < (Async & Abort[E])
+    )(using Frame): Fiber[E, A] < IO =
+        repeatWithDelay(Schedule.delay(startAfter).andThen(Schedule.fixed(delay)), state)(f)
+
+    def repeatWithDelay[E, S](delaySchedule: Schedule)(f: => Unit < (Async & Abort[E]))(using Frame): Fiber[E, Unit] < IO =
+        repeatWithDelay(delaySchedule, ())(_ => f)
+
+    def repeatWithDelay[E, A: Flat, S](
```

Comment:
Instead of relying on the underlying java timer to manage repeated tasks, this new `Timer` implementation spins up new fibers that use the provided `Schedule` to repeat and `Clock` to sleep when necessary. This approach also allows scheduled tasks to have internal state between iterations and return a final value at the end (see the `A` type).

Reply:
I think this is much more useful!